This change helps user to switch between master keys

sharidas opened the pull request:
and user specific keys.

Signed-off-by: Sujith H <sharidasan@owncloud.com>

<!--
Thanks for submitting a change to ownCloud!

This is the bug tracker for the Server component. Find other components at https://github.com/owncloud/core/blob/master/.github/CONTRIBUTING.md#guidelines

For fixing potential security issues please see https://owncloud.org/security/

To make it possible for us to get your change reviewed and merged please fill out below information carefully.

Please note that any kind of change first has to be submitted to the master branch which holds the next major version of ownCloud.

We will carefully discuss if your change can or has to be backported to stable branches.
-->

## Description
This change helps us to make master key as default for encryption instead of default configuration we had.

## Related Issue
<!--- This project only accepts pull requests related to open issues -->
<!--- If suggesting a new feature or change, please discuss it in an issue first -->
<!--- If fixing a bug, there should be an issue describing it with steps to reproduce -->
<!--- Please link to the issue here: -->
fixes https://github.com/owncloud/core/issues/26847

## Motivation and Context
<!--- Why is this change required? What problem does it solve? -->

## How Has This Been Tested?
<!--- Please describe in detail how you tested your changes. -->
<!--- Include details of your testing environment, and the tests you ran to -->
<!--- see how your change affects other areas of the code, etc. -->
Testing yet to be done. This has only UI change.

## Screenshots (if appropriate):

## Types of changes
<!--- What types of changes does your code introduce? Put an `x` in all the boxes that apply: -->
- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:
<!--- Go over all the following points, and put an `x` in all the boxes that apply. -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
- [x] My code follows the code style of this project.
- [x] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [ ] I have read the **CONTRIBUTING** document.
- [ ] I have added tests to cover my changes.
- [ ] All new and existing tests passed.

sharidas committed:
This change helps user to switch between master and user specific keys.

Signed-off-by: Sujith H <sharidasan@owncloud.com>

sharidas commented:
I would like to know if the UI part looks ok?

sharidas commented:
@felixheidecke Any suggestion?

felixheidecke commented:
@sharidas Could you provide a few screenshots? Would be the fastest way to get feedback ;-)

PVince81 commented on settings/js/admin.js:
```diff
@@ -41,6 +41,61 @@ $(document).ready(function(){
+		$('#confirm-encryption-warning').addClass("hidden");
+	}
+
+	if ($('#enableEncryption').prop("checked") == true) {
```
always use strict equality `===`.

please setup jshint/jslint, it will show you warnings for this

PVince81 commented on settings/js/admin.js:
```diff
@@ -41,6 +41,61 @@ $(document).ready(function(){
 	});
 
+	function selectEncryptionBehavior() {
+		$('#confirm-encryption-warning').addClass("hidden");
```
this code is difficult to read

I suggest to have only a single method, like "onChange" which then toggles the options based on the conditions.
For example:

`$('#encryptionAPI').toggleClass(!userSpecificEncryption, 'hidden')`

This way you only need a single row instead of having it appear three times in different functions

PVince81 commented on settings/js/admin.js:
```diff
@@ -41,6 +41,61 @@ $(document).ready(function(){
+	});
+
+	$('#reconfirm-encryption-type').click(function (event) {
+		console.log("I am clicked the value selected is = ", $('#keyTypeId option:selected').val());
```
remove logging

PVince81 commented on settings/js/admin.js:
```diff
@@ -41,6 +41,61 @@ $(document).ready(function(){
+		console.log("I am clicked the value selected is = ", $('#keyTypeId option:selected').val());
+		if ($('#keyTypeId option:selected').val() == 'masterkey') {
+			OC.AppConfig.setValue('encryption', 'useMasterKey', '1');
+			location.reload();
```
the function `OC.AppConfig.setValue()` does an ajax call, so if you reload the page directly after that there is a slight risk of race condition in which the value isn't completely saved. Usually when reloading the page the browser will abort all ajax connections.

only reload the page once the ajax call finished

PVince81 commented on settings/templates/panels/admin/encryption.php:
```diff
@@ -1,3 +1,18 @@
+	<select id="keyTypeId" name="keyType">
+		<option value="nokey">Please select an encryption option</option>
+		<option value="masterkey" <?php echo \OC::$server->getAppConfig()->getValue('encryption', 'useMasterKey', 0) !== 0 ? 'selected' : ''; ?>>Master Key</option>
+		<option value="customkey">User-specific key</option>
```
text translation missing, please recheck all the texts and make sure you're using the translation function

sharidas commented:
@PVince81 I have updated the patch set with the changes mentioned below:
1) Replaced the drop texts with the translation function
2) Used ajaxStop function after the setValue's used.
3) Removed console.log which was there in the code.
4) Used === for equality check.
5) Wrote a single function which gets called for the onChange event of the drop down selection.
6) Tried adding some comments, to make the code readable.

Let me know if the code looks better, else I can refactor again.

PVince81 commented on settings/js/admin.js:
```diff
@@ -41,6 +41,85 @@ $(document).ready(function(){
+			OC.AppConfig.setValue('core', 'encryption_enabled', 'yes');
+			OC.AppConfig.setValue("encryption", "encryptHomeStorage", '0');
+			OC.AppConfig.setValue('encryption', 'useMasterKey', '1');
+			$(document).ajaxStop(function () {
```
No, this will break in some cases.

Not only this code is doing ajax calls. There are background processes like the notifications app, the heartbeat, etc that will trigger ajax calls.

A better approach is to use promises:
- adjust `setValue` and other methods to return the result of `$.post` or other ajax calls
- use `$.when(call1, call2, call3).then(function() { ... })` to find out when all the three calls have finished.

PVince81 commented:
The flow is wrong, "Enable server-side encryption" must always be the first entry.
Then after that the selection of encryption type.

Also, remember that this checkbox "Enable server-side encryption" is about enabling the encryption subsystem/engine in core, not enabling a specific module.

The code with the encryption type selection needs to be part of the "Default encryption module" app (aka plugin). Keep in mind that other developers can write their own "Custom encryption module" (aka plugin) in which case the new dropdown is not relevant at all.

The encryption module code is in the "apps/encryption" folder. Please move the dropdown logic there.

PVince81 commented:
Flow as follows:

1. Admin enables the app "default encryption module"
2. Admin goes to admin page
3. Admin clicks "Enable server-side encryption" then "Enable encryption"
4. Admin scrolls down to the "Encryption" section (let's rename it to "Default encryption module")
5. Admin selects encryption type (which is the only visible item at this stage)
6. If admin picked "user-keys": now display all old fields
    - if admin picked "master-key": no further fields

The dropdown needs a button "Select this mode".
Once clicked, the dropdown must disappear and be replaced with a label: "Encryption type: xyz" which cannot be changed any more.

sharidas commented:
@PVince81 Thanks for the review comments. I have updated the PR and below are the changes made in the updated PR:
1) Made the work flow as follows:
    After enabling the encryption app and after re-login, admin when navigates to "encryption" page
    a) "Encryption" has been renamed to "Default encryption module"
    b) Once user selects either of the options: "Master Key" or "User-specific Key", a "select this mode" button appears
     c) Once the button is clicked based on the respective mode, the drop down and the button is not visible to the user. And a label appers "Encryption: master key" if master was selected or "Encryption: User specific key" if user specific key was selected.
     d) If user had opted for the master key then page reload happens
     e) The logic for the user specific key remains the same.

Let me know how the current PR looks?

PVince81 commented:
- [ ] BUG: Getting this message after "enable encryption" then "enable default encryption module app":
"Encryption App is enabled but your keys are not initialized, please log-out and log-in again"

This message shouldn't appear until an encryption type has been selected, because this message is only relevant for user-key mode encryption.

- [ ] TODO: change the button text to "Permanently select this mode"
- [ ] BUG: do not display anything under the dropdown until a mode is selected. So selecting "user-key" should not show anything either. Once the selection is made, the dropdown disappears forever and the rest of the form can appear.

PVince81 commented:
Enabling the modes worked fine, nice!

PVince81 commented on apps/encryption/js/settings-admin.js:
```diff
@@ -8,7 +8,77 @@
+
+	encryptionTypeSelection($("#keyTypeId :selected").val(), "static");
+
+	OC.AppConfig.getValue("encryption", "useMasterKey", function ($value) {
```
note: a way to prevent reading these values with ajax would be to append them into the template on the PHP side then read them out with JS.

PVince81 commented on apps/encryption/js/settings-admin.js:
```diff
@@ -8,7 +8,77 @@
+		//Action to be taken when "Select this mode" button is selected.
+		$("#select-mode,#keyTypeId").addClass("hidden");
+		if($("#keyTypeId :selected").val() === "masterkey") {
+			$.when(OC.AppConfig.setValue("encryption", "encryptHomeStorage", '0'),
```
this cannot work properly, you didn't adjust `OC.AppConfig.setValue` to actually return the ajax object

PVince81 replied:
well, the two "setValue" calls to work, but the final callback to reload the page is never called

sharidas commented:
@PVince81 I have updated the PR for review by making the changes as follows:
1) Fixed bug : Getting this message after "enable encryption" then "enable default encryption module app": "Encryption App is enabled but your keys are not initialized, please log-out and log-in again"
2) Changed button text to "Permanently select this mode"
3) Once the user clicks the "Permanently select this mode", the rest of the form is displayed and the drop down is not displayed for ever.
4) For the getValue thing, I have moved the call to the PHP template and the JS code reads from there.
5) I have made adjustments to call the  2 setValue and passed the promise objects to when. I did a verification using firebug. When page reload is called, the setValues are seen to be completed their job. Verified in the network response section of firebug.

PVince81 commented on apps/encryption/js/settings-admin.js:
```diff
@@ -8,7 +8,90 @@
+			//If user selects "Master Key" from the drop down
+			$("#select-mode").removeClass("hidden");
+
+			$("#encryptHomeStorageSetting, #encryptionSetRecoveryKey").addClass("hidden");
```
hide these permanently before any selection is made and only unhide if the other type is selected

PVince81 replied:
what about this comment ? this code isn't needed as you need to hide these fields only initially.

PVince81 replied:
@sharidas any update here ?

sharidas replied:
looking into.

sharidas replied:
I have removed this line. As its not required. Its toggled from hide only when user specific keys are permanently selected.

PVince81 commented on apps/encryption/js/settings-admin.js:
```diff
@@ -8,7 +8,90 @@
+			if(state === "static") {
+				$("#select-mode, #keyTypeId").addClass("hidden");
+				$("#encryptHomeStorage, #encryptionSetRecoveryKey").addClass("hidden");
+				if($("#encryptionType").text().trim().length === 0) {
```
why `$("#encryptionType").text()` and not simply the value of `encryptionType` above ?

PVince81 commented on apps/encryption/js/settings-admin.js:
```diff
@@ -8,7 +8,90 @@
+
+	//This function is to arrange the display of the encryption page
+	//in the settings.
+	function encryptionTypeSelection(encryptionType, state=undefined) {
```
add a proper JSDoc block, see in other JS files how it looks like, for example: https://github.com/owncloud/core/blob/stable9.1/core/js/js.js#L181

and explain what "state" means

PVince81 commented on apps/encryption/js/settings-admin.js:
```diff
@@ -8,7 +8,90 @@
+
+			}
+		} else {
+			//If user selects "Please sect an encryption option" from the drop down
```
s/sect/select/

PVince81 commented on apps/encryption/js/settings-admin.js:
```diff
@@ -8,7 +8,90 @@
+			function adjustAjaxCallMasterKey () {
+				return $.ajax({
+					success: function() {
+						return OC.AppConfig.setValue('encryption', 'useMasterKey', '1');
```
currently `setValue` never returns anything, please check the method declaration here: https://github.com/owncloud/core/blob/master/core/js/config.js#L40

That's where you need to add a `return` to make it return the ajax object.

PVince81 commented on apps/encryption/js/settings-admin.js:
```diff
@@ -8,7 +8,90 @@
+			var masterAjaxObj = adjustAjaxCallMasterKey();
+			var userSpecificAjaxObj = adjustAjaxCallUserKey();
+			$.when(masterAjaxObj, userSpecificAjaxObj).done(function (masterKeyObj, userSpecificObj) {
+				location.reload();
```
have you tested this ? did it work ?

from reading the code above this cannot work.
try adding `console.log('test')` and see if something is logged after setting the master key.

PVince81 replied:
Ok, so it seems to reload the page in my local test, but not for the right reason.

You made empty `$.ajax` calls that do nothing, so their promises/deferred will automatically be in the resolved state.
This is still wrong because it might happen before `setValue` finished.
Please use the approach I asked for above.

PVince81 commented on apps/encryption/lib/Session.php:
```diff
@@ -60,7 +60,10 @@ public function setStatus($status) {
 		if (is_null($status)) {
-			$status = self::NOT_INITIALIZED;
+			if(\OC::$server->getAppConfig()->getValue('encryption', 'useMasterKey', '0') !== '0'
+			  or \OC::$server->getAppConfig()->getValue('encryption', 'encryptHomeStorage', '') !== '') {
```
Ok, so you're using the value "encryptHomeStorage" to find out whether user-key encryption is enabled, is that correct ?

PVince81 commented on apps/encryption/templates/settings-admin.php:
```diff
@@ -5,19 +5,46 @@
+		<label id="encryptionType">
+			<?php
+			$masterKey = \OC::$server->getAppConfig()->getValue('encryption', 'useMasterKey', '0');
+			$homeEncryption = \OC::$server->getAppConfig()->getValue('encryption', 'encryptHomeStorage', 'empty');
```
In the above logic you used `''` as empty value, which is correct.
Why not use `''` as well instead of `'empty'` for consistency ?

Developers reading both pieces of code might get confused and think that "empty" is a different value than "".

PVince81 commented on apps/encryption/templates/settings-admin.php:
```diff
@@ -5,19 +5,46 @@
+		<p id="encryptionKeySelection">
+			<select id="keyTypeId" name="keyType">
+				<option value="nokey"><?php p($l->t("Please select an encryption option"))?></option>
+				<option value="masterkey" <?php echo \OC::$server->getAppConfig()->getValue('encryption', 'useMasterKey', '0') !== '0' ? 'selected="selected"' : ''; ?>><?php p($l->t("Master Key"))?></option>
```
Use `print_unescaped` instead of `echo`

PVince81 commented on apps/encryption/templates/settings-admin.php:
```diff
@@ -5,19 +5,46 @@
+				<option value="customkey"><?php p($l->t("User-specific key"))?></option>
+			</select>
+			<button id="select-mode" type="button" class="hidden"><?php p($l->t("Permanently select this mode"));?></button>
+			<label id="masterKeyVal" class="hidden"><?php echo \OC::$server->getAppConfig()->getValue("encryption", "useMasterKey", "empty"); ?> </label>
```
In the past we usually used hidden input fields to make server values accessible to the JS code.
Might be better than using label.

Another better approach is using data attributes on a specific element: `<div id="encryptionSettingsContainer" data-master-key="true">`

PVince81 commented:
- [ ] BUG: when enabling master key, the files are not encrypted at all (check any newly uploaded file under "data/admin/files/").

When I switch to master with the oc_appconfig settings from this branch, files also do not get encrypted.

However if I setup encryption from scratch on master with master key, it works.
So maybe something else is missing in the settings. Maybe compare the oc_appconfig values from both branches.

sharidas commented:
@PVince81 I have updated the PR for review with the changes mentioned below:
a) I have added return before the https://github.com/owncloud/core/blob/master/core/js/config.js#L40.
b) Instead of using labels with hidden, div is being used with proper attributes.
c) Changed the code flow for User Specific Key. Now oc_appconfig will have userSpecificKey in "encryption" app. This will be set only when user selects User Specific Key mode in the UI.
d) Replaced echo with print_unescaped where mentioned.
e) Removed 'empty' and used '' for consistency.
f) Verified with Master Key and User Specific Key the data is getting encrypted. Verified by uploading a text file after relogin. And then cat the file inside data folder. I was able to see message:
"
HBEGIN:oc_encryption_module:OC_DEFAULT_MODULE:cipher:AES-256-CTR:signed:true:HEND
"
at the beginning of the line.

Let me know if more changes are required.

PVince81 commented on apps/encryption/js/settings-admin.js:
```diff
@@ -8,7 +8,73 @@
+				location.reload();
+			});
+		} else if($("#keyTypeId :selected").val() === "customkey") {
+			OC.AppConfig.setValue("encryption", "userSpecificKey", '1');
```
so master key reloads the page but not this one ? any particular reason ?

PVince81 replied:
also, since you added a new key it is not guaranteed that upgrading from 9.1 with encryption already enabled will work because this key will be missing. Need to find a way to solve this.

sharidas replied:
my mistake.

PVince81 commented on core/js/config.js:
```diff
@@ -37,7 +37,7 @@ OC.AppConfig={
 	},
 	setValue:function(app,key,value){
-		OC.AppConfig.postCall('setValue',{app:app,key:key,value:value});
+		return OC.AppConfig.postCall('setValue',{app:app,key:key,value:value});
```
`postCall` doesn't return anything, please fix accordingly

sharidas commented:
@PVince81 I have updated the PR with the change to return from the postCall. I have verified the same using debugger: http://pasteboard.co/1MeQdMvX.png. I have also added the reload. Made sure that before the reload ( the second reload ) happens, the ajax request is completed. I have verified the data uploaded after relogin ( for both master key and user specific key) they found to be encrypted.

PVince81 commented:
Works fine now.

Now to some additional missing bits:
- [ ] TODO: add occ command to select encryption type and remove "occ encrpyiton:enable-master-key"
- [ ] TODO: update documentation (raise ticket in doc repo) to update the web UI flow and CLI flow (mark the PR as pending until this PR here is merged)

I think later on we'll mark "Master key" in as "recommended", but not in this PR because there are still many features that do not work well with master key, so I'm not sure.

PVince81 commented:
- [ ] TODO: adjust tests/integration/run.sh to use the new commands to enable either user-key or master-key mode

Note: these tests only run on the master branch so you'll need to run them locally.

For example:
```
% cd tests/integration
% sudo -uwwwrun OC_TEST_ENCRYPTION_ENABLED=1 OC_TEST_ENCRYPTION_MASTER_KEY_ENABLED=1 ./run.sh features/webdav-related.feature
```

PVince81 commented:
@sharidas would be good to finish this, see my points above

sharidas commented:
@PVince81 I have updated the patch. I have removed "encrpyiton:enable-master-key" and created "encryption:select-encryption-type". It would accept either "masterkey" or "user-keys" ( without quotes). And confirmation question is also taken care. Apart from that I have also looked into run.sh. I have tried to update by removing "encryption:enable-master-key" and using "encryption:select-encryption-type". Tried testing it from the command line by enabling the app, enabling encryption, selecting either of the type {masterkey|user-keys}. Logged out from the UI. Uploaded files. Verified the header of the files. And they were encrypted.

sharidas commented:
Umm. PhantomJS error!! Tried to shake the jenkins by pushing again. But its failing again :(

sharidas commented:
So below are the steps done for validating the feature:
- [x] Create an admin user, "admin"
- [x] The admin user enables "Default encryption module" app by navigating "Settings" -> "Apps" and click "Show disabled apps". And then click "Enable" button underneath the App icon.
- [x] Navigate from Apps to "Encryption" available in the right panel of the screen, under the "Admin" Section.
- [x] Now enable server-side encryption. Reconfirm by clicking the "Enable encryption" button.
- [x] Select one of the options, "Master Key", available under "Default Encryption Module". And click "Permanently select this mode".
- [x] Logout and login back to admin user.
- [x] Try to create a file hello.txt. And add some content to the text. Or try to upload a text.
- [x] Now cat the file under data/admin/files/hello.txt. You should be able to see that file has been encrypted.
- [x] Going further with the next test, do a scratch install of oC and then try to select option "User-specific key" and click "Permanently select this mode" button.
- [x]  Logout and login back as admin user
- [x] Try to upload a text file. When user tries to cat the file uploaded can see the data is encrypted.
- [x] The same result can be obtained when tried with a normal user. This holds true for verification with master key too.
- [x] Now try to add a recovery key.
- [x] Try to modify the password of the normal user created. It won't happen until "Admin Recovery Password" is provided.

Next task to verify the command line options
- [x] Execute the command : ./occ maintenance:install to do a fresh install of oC.
- [x] Execute the command ./occ app:enable encryption, to enable the app.
- [x] Execute the command ./occ encryption:enable, to enable encryption
- [x] Execute the command ./occ encryption:select-encryption-type masterkey. This will prompt a question which requires user input. If user is ok press 'y' ( without quotes ) to continue. User can see master key enabled message on the screen.
- [x] Now from the UI try to login as admin user and try to upload the text. Execute cat command on the text file to see the file has been encrypted.
- [x] The step before can be executed with ./occ encryption:select-encryption-type masterkey -y, this wont prompt the question and hence no user input required. To bypass the step of asking query to user and providing 'y' we directly provide -y option in the command. This would provide the same result when user tries to upload a file.

- [x] Execute the command : ./occ maintenance:install to do a fresh install of oC.
- [x] Execute the command ./occ app:enable encryption, to enable the app.
- [x] Execute the command ./occ encryption:enable, to enable encryption
- [x] Execute the command ./occ encryption:select-encryption-type user-keys which prompts a query to user that requires a user feedback as 'y'( without quotes ) or user can execute ./occ encryption:select-encryption-type user-keys -y.
- [x] Try to login back as admin user and then upload a file. cat the file and user can see file is encrypted.
- [x] Have also verified that once either of the options selected master key or user specific key, cannot be made any change from the UI. And even from the command line.
- [x] While the write up I have found a minor update. Which I have updated in the PR.

PVince81 commented:
- [ ] add spinner while encryption is being enabled. It took a few seconds to enable master key and it makes the UI look broken (check how other spinners are displayed for example in the share panel, there's a CSS class for it)
- [ ] the text "Encryption: master key" that appears after going to that page again after enabling, the text must not display a hand cursor as it's not clickable

Otherwise this looked good in my quick local tests.

sharidas commented on apps/encryption/js/settings-admin.js:
```diff
@@ -8,7 +8,77 @@
+	$("#select-mode").click(function () {
+		//Action to be taken when "Select this mode" button is selected.
+		$("#loadSpinner").toggleClass("hidden");
+		$("#loadSpinner").addClass("loading");
```
Does this spinner look ok?

PVince81 replied:
Argh, global ids again ? What if another section also has a spinenr called "loadSpinner" ?

Please use class names instead.
The only id that's acceptable here is the parent element `#encryptionKeySelection`, so use a selector like `#encryptionKeySelection .loading` when toggling or something.

And when toggling please force the value to "true" as you might not know what its previous state is.

sharidas commented on apps/encryption/js/settings-admin.js:
```diff
@@ -8,7 +8,76 @@
+
+	$("#select-mode").click(function () {
+		//Action to be taken when "Select this mode" button is selected.
+		$("#encryptionKeySelection").toggleClass("loading", true);
```
This looks better, right?

PVince81 replied:
Did you test it ?

The spinner only shortly appears at the very right of the row and disappears very quickly.

Also looking at the code it looks like it sets the class on the whole container, which looks wrong.
The spinner should have its own div element.

sharidas commented on apps/encryption/js/settings-admin.js:
```diff
@@ -8,7 +8,78 @@
+		//Action to be taken when "Select this mode" button is selected.
+		var $loadSpinner = $('#encryptionKeySelection').find('div.hidden').first();
+		$loadSpinner.toggleClass('hidden',false);
+		$loadSpinner.toggleClass('loading',true);
```
Made the modification by having the div for spinner. And now atleast the spinner doesn't vanish quickly.

PVince81 commented:
Looks good now 👍
